<html>
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no">

    <title>Sleepwalk</title>
    <link rel="stylesheet" href="../revealjs/dist/reset.css">
    <link rel="stylesheet" href="../revealjs/dist/reveal.css">
    <link rel="stylesheet" href="../revealjs/dist/theme/insomnia.css">
    <link rel="stylesheet" type="text/css" href="src/css/prism.css">
  </head>
  <body background-size="cover">
  	<div class="reveal">
  	  <div class="slides">
  	  	<section style="background-color: #000;">
  	  	  <div>
  	  	    <h1>Sleepwalk</h1>
  	  	    <h2>A Tool to Interactively Explore Dimension-Reduced Embeddings</h2>
  	  	  </div>
  	  	  <img src="src/img/insomnia.png" class="plain" width="15%" style="margin: 5px;">
  	  	  <table>
  	  	    <tr width="100%">
  	  	      <td style="color: #d4c78c;">Presented by:</td>
  	  	      <td style="color: #d2dae7;">Svetlana Ovchinnikova</td>
  	  	    </tr>
  	  	    <tr>
  	  	      <td style="color: #d4c78c;">Slides at:</td>
  	  	      <td style="color: #d2dae7;">kloivenn.github.io/CRC1366</td>
  	  	    </tr>
  	  	  </table>
  	  	  <div class="twosided">
  	  	    <small style="color: #d4c78c; text-align: center;">
  	  	      January 19th, 2022
  	  	    </small>
  	  	    <small style="color: #d4c78c; text-align: center;">
  	  	      <s>Asselheim</s>
  	  	    </small>  
  	  	  </div>
  	  	</section>
  	  	<section>
  	  	  <h2>Dimensionality Reduction</h2>
  	  	  <div class="content">
  	  	    <div class="content-pads">
  	  	      <span class="fragment fade-in-and-out"></span>
  	  	      <span class="fragment fade-in-and-out"></span>
  	  	      <div data-animate>
  	  	        <svg width="900" height="500">
  	  	          <g id="sc_table">
	  	  	          <image href="src/img/table.png" width="300" x="50" y="30"/>
	  	  	          <text x="200" y="260" text-anchor="middle">(m cells x n genes)</text>
	  	  	        </g>
  	  	          <image href="src/img/clusters.png" width="300" x="550" y="150"/>
  	  	          <g id="dimred_arrow">
  	  	            <image href="src/img/arrow_long.svg" width="200" x="380" y="160" transform="rotate(30, 380, 160)"/>                    
  	  	            <text text-anchor="middle" x="470" y="210" transform="rotate(30, 470, 210)">
  	  	              <tspan x="470">Dimensionality</tspan>
  	  	              <tspan x="470" dy="1.4em">reduction</tspan>
  	  	            </text>
  	  	          </g>
  	  	          <text x="700" y="140" text-anchor="middle">2D embedding</text>
  	  	          <g id="graph" transform="translate(150, 410)">
  	  	            <image href="src/img/arrow_long.svg" width="120" y="-5" x="-5"/>
  	  	            <image href="src/img/arrow_long.svg" width="120" y="-5" x="-5" transform="rotate(-90)"/>
  	  	            <image href="src/img/arrow_long.svg" width="120" y="-5" x="-5" transform="rotate(135)"/>
  	  	            <text font-size="70%" x="70" y="-10">gene 1</text>
  	  	            <text font-size="70%" x="70" y="-10" transform="rotate(-90)">gene 2</text>
  	  	            <text font-size="70%" x="70" y="-10" transform="rotate(-225) rotate(180,70,-10) translate(-40,-20)">gene 3</text>
  	  	            <circle r="5" cx="30" cy="-80"/>
  	  	            <circle r="5" cx="40" cy="-20"/>
  	  	            <circle r="5" cx="-30" cy="-10"/>
  	  	            <circle r="5" cx="10" cy="30"/>
  	  	            <circle r="5" cx="60" cy="50"/>
  	  	            <text x="40" y="85">n dimensional space</text>
  	  	          </g>
  	  	        </svg>
  	  	        <!--
  	  	          {
  	  	          "setup": [{
  	  	          "element": "#graph",
  	  	          "modifier": "attr",
  	  	          "parameters": [{"opacity": 0}]   
  	  	          },{
  	  	          "element": "#dimred_arrow",
  	  	          "modifier": "attr",
  	  	          "parameters": [{"opacity": 0}]   
  	  	          },{
  	  	          "element": "#sc_table",
  	  	          "modifier": "attr",
  	  	          "parameters": [{"opacity": 0}]   
  	  	          }],
  	  	          "animation": [
 	  	          [{
  	  	          "element": "#dimred_arrow",
  	  	          "modifier": "attr",
  	  	          "parameters": [{"opacity": 1}] 
  	  	          }, {
  	  	          "element": "#sc_table",
  	  	          "modifier": "attr",
  	  	          "parameters": [{"opacity": 1}]
  	  	          }
  	  	          ], [{
  	  	          "element": "#graph",
  	  	          "modifier": "attr",
  	  	          "parameters": [{"opacity": 1}] 
  	  	          }, {
  	  	          "element": "#dimred_arrow",
  	  	          "modifier": "rotate",
  	  	          "parameters": [-50, 470, 210]
  	  	          }, {
  	  	          "element": "#dimred_arrow",
  	  	          "modifier": "translate",
  	  	          "parameters": [-40, 100]
  	  	          }
  	  	          ], []
  	  	          ]
  	  	          }                
  	  	        -->                   
  	  	      </div>
  	  	    </div>
  	  	  </div>
  	  	</section>
  	  	<section>
  	  	  <h2>Distortion Is Unavoidable</h2>
  	  	  <div class="content">
  	  	    <div class="content-pads">
  	  	      <img data-src="src/img/1ev1X.png">
  	  	    </div>
  	  	  </div>
  	  	  <p><a target="about:blank" href="https://xkcd.com/977/">https://xkcd.com/977/</a></p>
  	  	</section>
  	  	<section>
  	  	  <h2>Sleepwalk</h2>
  	  	  <div class="content">
  	  	    <div class="content-pads">
  	  	      <pre><code data-trim data-noescape class="language-r">sleepwalk(citeSeq$tsne, features)</code></pre>
  	  	      <div class="aspect-ratio" style="padding-top: 45%; width: 75%; margin-left: 10%;">
  	  	        <iframe data-src="src/Fig_B.html"></iframe>
  	  	      </div>
  	  	    </div>
  	  	  </div>
  	  	  <div class="wrapper" style="z-index: -1;">
  	  	    <div class="bottom-cell">
  	  	      <p class="footnote">data from 
  	  	        <a target="about:blank" href="https://doi.org/10.1038/nmeth.4380">Stoeckius et al., Simultaneous epitope and transcriptome measurement in single cells, Nature Methods, 14:865, 2017</a>
  	  	      </p>
  	  	    </div>
  	  	  </div>
  	  	</section>
  	  	<section>
  	  		<h2>How to run</h2>
  	  		<div class="content">
  	  			<div class="content-pads">
  	  				<pre><code data-trim data-noescape class="language-r">library(Matrix)
library(irlba)
library(uwot)
library(sleepwalk)

rawCounts <- readRDS("citeSeq_RNA.rds")

# calculate library sizes (size factors) and normalize data
sf <- colSums(rawCounts)
normCounts <- t(t(rawCounts/sf * 10000))

# get PCA and UMAP
pca <- prcomp_irlba(log1p(t(normCounts)), n = 50)
um <- umap(pca$x, min_dist = 0.3, n_neighbors = 30, spread = 3)

# run Sleepwalk
sleepwalk(um, pca$x)</code></pre>					
  	  			</div>
  	  		</div>
  	  		<div class="wrapper" style="z-index: -1;">
  	  		  <div class="bottom-cell">
  	  		  	<p class="footnote">download data:  
  	  		  	  <a target="about:blank" href="R/citeSeq_RNA.rds">citeSeq_RNA.rds</a>
  	  		  	</p>
  	  		  	<p class="footnote">download data:  
  	  		  	  <a target="about:blank" href="R/citeSeq_ADT.rds">citeSeq_ADT.rds</a>
  	  		  	</p>  	  		  	
  	  		  	<p class="footnote">data from 
  	  		  	  <a target="about:blank" href="https://doi.org/10.1038/nmeth.4380">Stoeckius et al., Simultaneous epitope and transcriptome measurement in single cells, Nature Methods, 14:865, 2017</a>
  	  		  	</p>
  	  		  </div>
  	  		</div>  	  		
  	  	</section>
  	  	<section>
  	  		<h2>For Seurat Users</h2>
  	  	  <div class="content">
  	  	    <div class="content-pads">
  	  	      <pre><code data-trim data-noescape class="language-r">sleepwalk(Embeddings(Reductions(seu, "umap")), Embeddings(Reductions(seu, "pca")))</code></pre>
  	  	      <div class="aspect-ratio" style="padding-top: 50%; width: 45%; margin-left: 25%;">
  	  	        <iframe data-src="src/cerebellum_lukas.html"></iframe>
  	  	      </div>
  	  	    </div>  	  		
  	  	</section>
  	  	<section>
  	  		<h2>Metric effect (code)</h2>
  	  		  	  		<div class="content">
  	  		  	  			<div class="content-pads">
  	  		  	  				<pre><code data-trim data-noescape class="language-r">library(sparseMatrixStats)
library(scales)
library(rlc)

# getting variable genes
means <- rowMeans(normCounts)
vars <- rowVars(normCounts)
lc_scatter(x = means, y = vars/means, logScaleX = 10, logScaleY = 10, size = 2,
           opacity = 0.4)
genes <- getMarked()

# scaling
variableGenes <- as.matrix(log1p(normCounts[genes, ]))
scaled <- (variableGenes - rowMeans(variableGenes))/rowSds(variableGenes)
scaled <- squish(scaled, c(-10, 10))

pca_scaled <- prcomp_irlba(t(scaled), n = 50)
um_scaled <- umap(pca_scaled$x, min_dist = 0.3,
                  spread = 3,
                  n_neighbors = 30, metric = "cosine")

sleepwalk(list(um, um_scaled), list(pca$x, pca_scaled$x), metric = c("euclid", "cosine"))</code></pre>
  	  		  	  			</div>
  	  		  	  		</div>  	  		
  	  	</section>
  	  	<section>
  	  		<h2>Metric effect (example)</h2>
  	  		<div class="content">
  	  		  <div class="content-pads">
  	  		    <div class="aspect-ratio" style="padding-top: 55%;">
  	  		      <iframe data-src="src/scaled.html"></iframe>
  	  		    </div>
  	  		  </div>  	
  	  	</section>
  	  	<section>
  	  		<h2>Ground truth</h2>
  	  		<div class="content">
  	  		  <div class="content-pads">
  	  		    <pre><code data-trim data-noescape class="language-r">countsADT <- readRDS("citeSeq_ADT.rds")
countsADT[1:10, 1:10]
normCountsADT <- t(countsADT)/colSums(countsADT) * 10000
sleepwalk(list(um, um_scaled), log1p(normCountsADT))</code></pre>
  	  		    <div class="aspect-ratio" style="padding-top: 45%; width: 75%; margin-left: 10%;">
  	  		      <iframe data-src="src/groundTruth.html"></iframe>
  	  		    </div>
  	  		  </div>
  	  		</div>
  	  	</section>
  	  	<section>
  	  		<h2>Comparing samples</h2>
            <div class="content">
              <div class="content-pads">
                <pre><code data-trim data-noescape class="language-r">sleepwalk(list(um13_A, um13_B, um14), 
      list(comFeatures13_A, comFeatures13_B, comFeatures14), same = "features",
      titles = c("e13.5_A", "e13.5_B", "e14.5"), nrow = 1)</code></pre>
                <div class="aspect-ratio" style="padding-top: 42%; width: 95%;">
                  <iframe data-src="src/Fig_D.html"></iframe>
                </div>
              </div>
            </div>
            <div class="wrapper" style="z-index: -1;">
              <div class="bottom-cell">
                <p class="footnote">data from 
                  <a target="about:blank" href="https://doi.org/10.1016/j.cub.2018.07.062">Carter et al., A Single-Cell Transcriptional Atlas of the Developing Murine Cerebellum, Current Biology, 28:18, 2018</a>
                </p>
              </div>
            </div>
  	  	</section>
  	  	<section>
  	  		<h2>Comparing samples (code)</h2>
  	  		  	  		<div class="content">
  	  		  	  			<div class="content-pads">
  	  		  	  				<pre><code data-trim data-noescape class="language-r">library(irlba)
library(uwot)
library(sleepwalk)

load("cerebellum_timepoints.RData")

getEmbedding <- function(data) {
  pca <- prcomp_irlba(data, n = 50)
  umap(pca$x, spread = 5, n_sgd_threads = 1, negative_sample_rate = 3)
}

um13_A <- getEmbedding(e13_A_norm)
um13_B <- getEmbedding(e13_B_norm)
um14 <- getEmbedding(e14_norm)

commonGenes <- intersect(intersect(colnames(e13_A_norm), colnames(e13_B_norm)), colnames(e14_norm))

pca <- prcomp_irlba(rbind(e13_A_norm[, commonGenes], e13_B_norm[, commonGenes], 
                          e14_norm[, commonGenes]), n = 50)
comFeatures13_A <- e13_A_norm[, commonGenes] %*% pca$rotation
comFeatures13_B <- e13_B_norm[, commonGenes] %*% pca$rotation
comFeatures14 <- e14_norm[, commonGenes] %*% pca$rotation

sleepwalk(list(um13_A, um13_B, um14), 
          list(comFeatures13_A, comFeatures13_B, comFeatures14), 
          same = "features", titles = c("e13.5_A", "e13.5_B", "e14.5"), nrow = 1)</code></pre>
  	  		  	  			</div>
  	  		  	  		</div>
   	  		<div class="wrapper" style="z-index: -1;">
  	  		  <div class="bottom-cell">
  	  		  	<p class="footnote">download data:  
  	  		  	  <a target="about:blank" href="R/cerebellum_timepoints.RData">cerebellum_timepoints.RData</a>
  	  		  	</p>
  	  		  </div>
  	  		</div>
  	  	</section>
  	  	<section>
  	  		<h2>Lasso selection</h2>
  	  	</section>
			</div>
		</div>
		<script type="text/javascript" src="src/js/prism.js"></script>
		<script src="../revealjs/dist/reveal.js"></script>
		<script src='../revealjs/plugin/notes/notes.js'></script>
		<script src="../revealjs/plugin/animate/plugin.js"></script>
		<script src="../revealjs/plugin/animate/svg.min.js"></script>

		<script>
		  // More info about config & dependencies:
		  // - https://github.com/hakimel/reveal.js#configuration
		  // - https://github.com/hakimel/reveal.js#dependencies
		  Reveal.initialize({
		    center: false,
		    hash: true,
		    slideNumber: 'c/t',
		    math: {
		      mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js',
		      config: 'TeX-AMS_HTML-full',
		      // pass other options into `MathJax.Hub.Config()`
		      TeX: { Macros: { RR: "{\\bf R}" } }
		    },

		    plugins: [ RevealNotes, RevealAnimate ]
		  });
		</script>		  	    	
  </body>
</html>
