<head>
	<link rel="stylesheet" type="text/css" href="src/linked-charts.css">
	<link rel="stylesheet" type="text/css" href="src/style.css">
	<script type="text/javascript" src = "src/linked-charts.min.js"></script>
	<script type="text/javascript" src = "src/data.js"></script>
	<title>Ex2plorer</title>
</head>
<body>
<div style="border-bottom: 1px solid #CCC;">
	<div id = "topbar">
		<img src = "src/Mammalian_evolutionTransparent.png" width = "150px" style = "margin-left: 10px;">
		<div><span style="font-style: italic; color: #B2303A">Ex<sup>2</sup></span>plorer</div>
	</div>
</div>
<div id = "lists">
	Compare
	<select id = "sp1" onchange = "changeSp1(this.value);">
		<option value = "Human">Human</option>
		<option value = "Mouse">Mouse</option>
		<option value = "Macaque">Macaque</option>
	</select>
	to
	<select id = "sp2" onchange="changeSp2(this.value);"></select>
	Tissue:
	<select id = "tissue" onchange="changeTissue(this.value);"></select>
</div>
	<table>
		<tr>
			<td id = "cross"></td>
			<td id = "lfs"></td> 
		</tr>
	</table>
	<table>
		<tr>
			<td>
				<div id = "genes">
				<p id = "notFound" style="color: red; visibility: hidden; text-align: center;">Gene with this ID was not found.</p>
				<span>Enter gene name:</span> <input id = "geneName" type = "text" onchange = "searchGene(this.value);"> <br>
				<span>Enter Ensembl ID:</span> <input id = "ensembl" type = "text" onchange = "searchEnsembl(this.value);">
				<table id = "info"></table>
				Search for this gene on <a target="_blank" rel="noopener noreferrer">GeneCards</a>.
				</div>
				<div class = "note1">
					Note: this database currently covers 6,327 1:1 orthologs shared across all six species studied in Wang et al. 2018.
			</div>
			</td>
			<td id = "expr"></td> 
		</tr>
	</table>
	<div style="text-align: right; font-size: 10px; color: #666; font-family: Qicksand, sans-serif; width: 1500px">
		Generated with the 
		<a target="_blank" rel="noopener noreferrer" href="https://anders-biostat.github.io/linked-charts/">LinkedCharts</a> library.
	</div>
	<script type="text/javascript">
		//var sp1 = "Human", sp2 = "Macaque", tissue = "Brain";
		var geneHover = 0;
		changeSp1("Human");

		var cross = lc.scatter("points")
			.x(function(i) {
				var val = lfData[sp1][sp2][tissue][0][i],
					domain = this.chart.get_domainX();
				if(val < domain[0])
					val = domain[0];
				if(val > domain[1])
					val = domain[1];
				return val;
			})
			.y(function(i) {
				var val = lfData[sp1][sp2][tissue][1][i]/lfData[sp1][sp2][tissue][0][i] - 1,
					domain = this.chart.get_domainY();
				if(val < domain[0])
					val = domain[0];
				if(val > domain[1])
					val = domain[1];
				return val;				
			})
			.colourValue(function(i) {
				var col = lfData[sp1][sp2][tissue][2][i];
				if(col == 0 || col == 2)
					return "Not significant";
				if(col == 3)
					return "Sig. Reinforcement";
				if(col == 1)
					return "Sig. Compensation";
			})
			.colourDomain(["Sig. Reinforcement", "Sig. Compensation", "Not significant"])
			.colourLegendTitle("")
			.palette(["#00C3C3", "#B2303A", "#aaa"])
			.size(function(i) {
				return (this.get_colourValue(i) == "Not significant") ? 2 : 4;
			})
			.opacity(function(i) {
				return (this.get_colourValue(i) == "Not significant") ? 0.5 : 0.75;
			})
			.domainX([-10, 10])
			.domainY([-5, 5])
			.axisTitleX("log2-fold change transcription")
			.axisTitleY("Translational Tuning Index")
			.informText(i => "<b>" + geneInfo[sp1][1][i] + "</b><br>" +
													"Chr: " + geneInfo[sp1][2][i] + "/" + geneInfo[sp2][2][i])
			.transitionDuration(0)
			.on_click(selectGene)
			.on_mouseover(moveErrorBars);

		var lfs = lc.scatter("points")
			.x(i => lfData[sp1][sp2][tissue][0][i])
			.y(i => lfData[sp1][sp2][tissue][1][i])
			.addColourScaleToLegend(false)
			.colourValue(function(i) {
				var col = lfData[sp1][sp2][tissue][2][i];
				if(col == 0 || col == 2)
					return "none";
				if(col == 3)
					return "Sig.R.";
				if(col == 1)
					return "Sig.C.";
			})
			.colourDomain(["none", "Sig.R.", "Sig.C."])
			.palette(["#aaa", "#00C3C3", "#B2303A"])
			.size(function(i) {
				return (this.get_colourValue(i) == "none") ? 2 : 4;
			})
			.opacity(function(i) {
				return (this.get_colourValue(i) == "none") ? 0.5 : 0.75;
			})
			.axisTitleX("log2-fold change transcription")
			.axisTitleY("log2-fold change translation")
			.informText(i => "<b>" + geneInfo[sp1][1][i] + "</b><br>" +
													"Chr: " + geneInfo[sp1][2][i] + "/" + geneInfo[sp2][2][i])
			.transitionDuration(0)
			.on_click(selectGene)			
			.on_mouseover(moveErrorBars);

		function selectGene(i) {
			selGene = i;
			expr.update();
			updateTable();
			d3.select("#ensembl").node().value = geneInfo[sp1][0][selGene];
			d3.select("#geneName").node().value = geneInfo[sp1][1][selGene];
			d3.select("#notFound")
				.style("visibility", "hidden");
		}

		lc.pointLine("error", cross)
			.nelements(2)
			.nsteps(2)
			.x(function(t, d) {
				var point = cross.get_layer("points").get_x(geneHover),
					se = lfData[sp1][sp2][tissue][4][geneHover];
				if(lc.isNaN(se)) se = cross.get_domainX()[1] - cross.get_domainX()[0];
				if(lc.isNaN(point)) {
					point = 0;
					se = 0;
				}

				if(d == 1)
					return point;
				if(t == 0)
					return point - se
				else
					return point + se;
			})
			.y(function(t, d) {
				var point = cross.get_layer("points").get_y(geneHover),
					se = lfData[sp1][sp2][tissue][3][geneHover];
				if(lc.isNaN(se)) se = cross.get_domainY()[1] - cross.get_domainY()[0];
				if(lc.isNaN(point)) {
					point = 0;
					se = 0;
				}

				if(d == 0)
					return point;
				if(t == 0)
					return point + se
				else
					return point - se;
			})
			.colour("black")
			.on_mouseover(function() {})
			.place("#cross");

		lc.pointLine("error", lfs)
			.nelements(2)
			.nsteps(2)
			.x(function(t, d) {
				var point = lfs.get_layer("points").get_x(geneHover),
					se = lfData[sp1][sp2][tissue][4][geneHover];
				if(lc.isNaN(se)) se = cross.get_domainX()[1] - cross.get_domainX()[0];
				if(lc.isNaN(point)) {
					point = 0;
					se = 0;
				}

				if(d == 1)
					return point;
				if(t == 0)
					return point - se
				else
					return point + se;
			})
			.y(function(t, d) {
				var point = lfs.get_layer("points").get_y(geneHover),
					se = lfData[sp1][sp2][tissue][5][geneHover];
				if(lc.isNaN(se)) se = cross.get_domainY()[1] - cross.get_domainY()[0];
				if(lc.isNaN(point)) {
					point = 0;
					se = 0;
				}

				if(d == 0)
					return point;
				if(t == 0)
					return point + se
				else
					return point - se;
			})			
			.colour("black")
			.on_mouseover(function() {})
			.place("#lfs");

		lfs.get_layer("error").findElements = function() {return [];}
		cross.get_layer("error").findElements = function() {return [];}

		function moveErrorBars(i) {
			geneHover = i;
			cross.get_layer("error").update();
			lfs.get_layer("error").update();
			if(d3.event.path[7].id == "cross")
				showLabel(cross, d3.event);
			if(d3.event.path[7].id == "lfs")
				showLabel(lfs, d3.event);
		}

		function showLabel(chart, e) {
	    var rect = chart.container.node().getBoundingClientRect(),
      pos = [d3.max([d3.min([e.clientX - rect.left, chart.plotWidth()]), 0]), 
            d3.max([d3.min([e.clientY - rect.top, chart.plotHeight()]), 0])]; 
    	chart.container.selectAll(".inform")
        .style("left", (pos[0] + 10) + "px")
        .style("top", (pos[1] + 10) + "px")
        .classed("hidden", false)
				.select(".value")
          .html(chart.get_layer("points").get_informText(geneHover));  
		}

		selGene = 0;
		d3.select("#ensembl").node().value = geneInfo[sp1][0][selGene];
		d3.select("#geneName").node().value = geneInfo[sp1][1][selGene];

		var expr = lc.scatter()
			.elementIds(() => d3.range(sampleTable.tissue.length).filter(i => sampleTable.tissue[i] == tissue))
			.width(800)
			.height(350)
			.set_margins({bottom: 90})
			.symbolLegendTitle("Species")
			.colourLegendTitle("")
			.symbolValue(i => sampleTable.species[i])
			.palette(["#000", "#aaa"])
			.x(i => sampleTable.species[i] + "_" + sampleTable.rep[i])
			.y(i => geneExpr[i][selGene])
			.ticksRotateX(45)
			.logScaleY(2)
			.axisTitleY("FPKM")
			.colourValue(i => sampleTable.type[i])
			//.colourValue(i => sampleTable.rep[i])
			.label(i => sampleTable.sampleId[i])
			.title(() => geneInfo[sp1][1][selGene])
			.place("#expr");


		var species = ["Human", "Mouse", "Macaque", "Opossum", "Platypus"];
		var cols = ["Ensembl ID", "Gene name", "Chr."],
			//rows = ["names"].concat(expr.get_domainX());
			rows = ["names"].concat(species);

		d3.select("#info")
			.selectAll("tr")
			.data(rows)
			.enter()
				.append("tr");
		d3.select("#info")
			.selectAll("tr")
				.selectAll("td")
				.data(d => ["names"].concat(d3.range(3)).map(e => [d, e]))
				.enter()
					.append("td");

		function updateTable() {
			d3.select("#info")
				.selectAll("td")
					.html(function(d) {
						if(d[0] == "names" && d[1] == "names")
							return "";
						if(d[0] == "names")
							return "<b>" + cols[d[1]] + "</b>";
						if(d[1] == "names")
							return "<b>" + d[0] + "</b>";
						if(d[1] == 1)
							return "<i>" + geneInfo[d[0]][d[1]][selGene] + "</i>";
						return geneInfo[d[0]][d[1]][selGene]; 
					})
			d3.select("#genes")
				.select("a")
					.attr("href", "https://www.genecards.org/cgi-bin/carddisp.pl?gene=" + geneInfo.Human[1][selGene]);
		}

		updateTable();

		function searchEnsembl(name) {
			d3.select("#notFound")
				.style("visibility", "hidden");			

			var sp;
			if(name.substring(0, 4) == "ENSG")
				sp = "Human";
			if(name.substring(0, 7) == "ENSMUSG")
				sp = "Mouse";
			if(name.substring(0, 7) == "ENSMMUG")
				sp = "Macaque";
			if(name.substring(0, 7) == "ENSMODG")
				sp = "Opossum";
			if(name.substring(0, 7) == "ENSOANG")
				sp = "Platypus";

			if(!sp){
				d3.select("#notFound")
					.style("visibility", "visible")
				return;
			}

			var gene = geneInfo[sp][0].indexOf(name);
			if(gene == -1) 
				d3.select("#notFound")
					.style("visibility", "visible")
			else {
				selGene = gene;
				updateTable();
				expr.update();
				d3.select("#geneName").node().value = geneInfo[sp1][1][selGene];
				moveErrorBars(gene);
			}
		}

		function searchGene(name) {
			d3.select("#notFound")
				.style("visibility", "hidden");			

			var gene = -1, i = 0;
			while(gene == -1 && i < species.length) {
				gene = geneInfo[species[i]][1].indexOf(name);
				i++;
			}
			if(gene == -1)
				d3.select("#notFound")
					.style("visibility", "visible")
			else {
				selGene = gene;
				updateTable();
				expr.update();
				d3.select("#ensembl").node().value = geneInfo[sp1][0][selGene];
				moveErrorBars(gene);				
			}
		}

		function changeSp1(name) {
			sp1 = name;
			var opts = d3.select("#sp2")
				.selectAll("option")
					.data(Object.keys(lfData[sp1]));
			opts.exit().remove();
			opts.enter()
				.append("option")
				.attr("selected", d => (!lfs && d == "Mouse") ? true : undefined)
				.merge(opts)
					.attr("value", d => d)
					.text(d => d);

			if(Object.keys(lfData[sp1]).indexOf(sp2) != -1)
				d3.select("#sp2").node().value = sp2
			else
				sp2 = d3.select("#sp2").node().value;

			changeSp2(sp2);
		}

		function changeSp2(name) {
			sp2 = name;
			var opts = d3.select("#tissue")
				.selectAll("option")
					.data(Object.keys(lfData[sp1][sp2]));
			opts.exit().remove();
			opts.enter()
				.append("option")
				.merge(opts)
					.attr("value", d => d)
					.text(d => d);

			if(Object.keys(lfData[sp1][sp2]).indexOf(tissue) != -1)
				d3.select("#tissue").node().value = tissue
			else
				tissue = d3.select("#tissue").node().value;

			changeTissue(tissue);
		}

		function changeTissue(name) {
			tissue = name;
			if(cross) {
				cross.update();
				cross.container
					.select("#legend")
						.select("#error")
							.select(".sample")
								.append("line")
									.attr("x1", 10)
									.attr("x2", 10)
									.attr("y1", 0)
									.attr("y2", 20)
									.attr("stroke", "black");
			}
			if(lfs) lfs.update();
			if(expr) expr.update();
		}

		expr.container
			.append("div")
				.attr("class", "note")
				.text("Note: the transcriptome and translatome data were normalized separately. " +
					"Thus, expression level values can only be compared across samples from the same data type.")

		d3.selectAll(".sample").selectAll("text")
			.attr("dy", 3);

		cross.legend.set_title({"error": ""});
		cross.legend.add_block([["Standard error"], [0]], "dash", "error");
						cross.container
					.select("#legend")
						.select("#error")
							.select(".sample")
								.append("line")
									.attr("x1", 10)
									.attr("x2", 10)
									.attr("y1", 0)
									.attr("y2", 20)
									.attr("stroke", "black");
	</script>
</body>